apps/ca.c

```diff
@@ -1697,8 +1697,7 @@ static int do_body(X509 **xret, EVP_PKEY *pkey, X509 *x509,
         while ((i = X509_NAME_get_index_by_NID(dn_subject,
                                                NID_pkcs9_emailAddress,
                                                -1)) >= 0) {
```

Comment:
Extra point for i = -1 prior while and then using i as last argument?

Reply:
Does this make sense? When you later call X509_NAME_delete_entry() won't it mess up the iteration? Your `i` won't be pointing to the same thing anymore right?

Reply:
Hmmm... Original rationale was that you restart search from same spot, so that if you remove entry, you start from one that was next to deleted. However! To my surprise loop in get_index starts with lastpos++. This ought to mean that it would remain blind to first element if you pass 0... And it would mean that it would skip over element next to deleted, i.e. miss it... So that correct solution would be initialize i to 0 and pass `i - 1`, not i... One can argue that reverting to original -1 would make code less controversial. Well, it of course depends in what's controversial, algorithmic inefficiency or avoiding controversy :-) I'm removing "pending 2nd review" since it has to be resolved... Thanks for heads up!

Reply:
Fixed , in a small different way